00-programming-fundamentals/git-03-diff-pull-and-merge.md

```diff
@@ -0,0 +1,251 @@
+
+With no arguments, `git show` displays the most recent commit
+
+```bash
+$ git commit -m "Added file.txt to test out git diff"
+$ git show
+```
+
+`git show` can display a commit other than the most recent one, but it takes a little work. We'll need to copy the **commit hash**, that long string of numbers and letters, that corresponds to our commit. You can find the commit hash using `git log`. Go ahead and copy it to your clipboard.
+
+![Finding a commit in the git log](images/git-log-output.png)
+
+Note that your commit hash will look different than the one in the image. To show that commit use `git show <paste-commit-hash-here>`
+
+```bash
+$ git show 157d556f2635b3c8fe838b3fa456af5145b77bae
+```
+
+## Collaborating with Git
+
+We've talked about using `git push` to send local changes to GitHub, but when using git to collaborate with others it's likely that you'll end up in the opposite situation, where there are changes on GitHub you don't have locally. Git provides the `git pull` command to handle this.
+
+To demonstrate `git pull`, we've provided a repository to work with. Follow these instructions to get started:
+
+1. Find a partner
+1. **One** partner should fork [this repository](https://github.com/Ada-C7/git-pull-activity).
+1. Whoever forked the repo should add the other partner as a collaborator.
+1. **Both** partners should clone the forked repo.
+1. **Both** partners should open `pull_practice.rb` in Atom.
+
+### Pulling Changes From GitHub
+
+Look at `Task 1` in the `pull_practice.rb`.
+
+1. **Partner 1** should uncomment the `duck_noise` method and save the file.
+1. **Partner 1** should then `git add` the file, `git commit`, and `git push`. You should now be able to see the changes on GitHub.
+1. **Partner 2** should run `git pull`. This will retrieve all commits from GitHub that aren't yet on the local repository.
+1. You should now be able to see the `duck_noise` method uncommented in partner 2's copy of the repo.
+
+Switch roles and do the same thing for `Task 2`, the `truck_noise` method.
+
+Running `git pull` when you don't have any changes is also known as performing a _fast-forward merge_.
+
+### Merging Changes
+
+What happens if both partners have made changes? Turns out git is pretty intelligent about putting together work from different sources. Most of the time, even if you're working in the same file, it can figure out the right thing to do.
+
+Turn your attention to `Task 3` in the activity. Here are your instructions:
+
+1. **Partner 1** should uncomment the `robot_noise` method and save the file.
+1. **Partner 1** should then `git add` the file, `git commit`, and `git push`. You should now be able to see the changes on GitHub.
+1. **Without pulling**, **Partner 2** should now uncomment the `train_noise` method and save the file.
+1. **Partner 2** should then `git add` the file and `git commit`.
+1. If **Partner 2** tries to `git push`, they should see an error message something like this:
+
+  ```
```

Comment:
I tried it and it ended up mistakenly syntax highlighting Git's output. I thought it was easier to read without.